myia/api.py

```diff
@@ -72,44 +56,250 @@ def _convert_function(env, fn):
+        except (TypeError, KeyError):
+            pass
+
+        return self.converters[type(value)](self, value)
+
+
+class Parser(PipelineStep):
+    """Pipeline step to parse a function.
+
+    Inputs:
+        input: A function.
+
+    Outputs:
+        graph: A graph.
+    """
+
+    def step(self, input):
+        """Assert that input is a Graph, and set it as the 'graph' key."""
+        g = self.resources.convert(input)
+        assert isinstance(g, Graph)
+        return {'graph': g}
+
+
+class Optimizer(PipelineStep):
+    """Pipeline step to optimize a graph.
+
+    Inputs:
+        graph: The graph to optimize.
+
+    Outputs:
+        graph: The optimized graph.
+    """
+
+    def __init__(self, pipeline_init, opts):
+        """Initialize an Optimizer."""
+        super().__init__(pipeline_init)
+        self.opts = opts
+
+    def step(self, graph):
+        """Optimize the graph using the given patterns."""
+        eq = PatternEquilibriumOptimizer(*self.opts, optimizer=self)
+        eq(graph)
+        return {'graph': graph}
+
+
+class Inferrer(PipelineStep):
+    """Pipeline step to run type/shape/value/etc. inference.
+
+    Inputs:
+        graph: The graph to infer.
+        argspec: Information about argument types.
+
+    Outputs:
+        inference_results: Inference results for the graph's output.
+        inferrer: The inference engine.
+    """
+
+    def __init__(self, pipeline_init, tracks, required_tracks, timeout):
+        """Initialize an Inferrer."""
+        super().__init__(pipeline_init)
+        self.tracks = tracks
+        self.required_tracks = required_tracks
+        self.timeout = timeout
+
+    def step(self, graph, argspec):
+        """Infer types, shapes, values, etc. for the graph."""
+        argprops = argspec
+        engine = InferenceEngine(
+            self.pipeline,
+            graph, argprops,
+            tracks=self.tracks,
+            required_tracks=self.required_tracks,
+            timeout=self.timeout
+        )
+        return {'inference_results': engine.output_info(),
+                'inferrer': engine}
+
+
+class Specializer(PipelineStep):
+    """Pipeline step to specialize a graph.
+
+    Inputs:
+        graph: The graph to specialize.
+        inferrer: The inference engine.
+
+    Outputs:
+        graph: The specialized graph.
+    """
+
+    def step(self, graph, inferrer):
+        """Specialize the graph according to argument types."""
+        spc = TypeSpecializer(inferrer)
+        result = spc.result
+        return {'graph': result}
+
+
+class ClosureConverter(PipelineStep):
+    """Pipeline step to closure convert a graph.
+
+    Inputs:
+        graph: The graph to closure convert.
+
+    Outputs:
+        graph: The closure converted graph.
+    """
+
+    def step(self, graph):
+        """Closure convert the graph."""
+        closure_convert(graph)
+        return {'graph': graph}
+
+
+class Exporter(PipelineStep):
```

Comment:
Might want to rename that to indicate that it runs with the DebugVM

Reply:
OK.